src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: BSD-2-Clause
+package org.xbill.DNS.lookup;
+
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+import static java.util.concurrent.CompletableFuture.*;
+import static java.util.stream.Collectors.toList;
+import static org.xbill.DNS.Type.CNAME;
+import static org.xbill.DNS.Type.DNAME;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Singular;
+import org.xbill.DNS.CNAMERecord;
+import org.xbill.DNS.Cache;
+import org.xbill.DNS.Credibility;
+import org.xbill.DNS.DClass;
+import org.xbill.DNS.DNAMERecord;
+import org.xbill.DNS.Message;
+import org.xbill.DNS.Name;
+import org.xbill.DNS.NameTooLongException;
+import org.xbill.DNS.Rcode;
+import org.xbill.DNS.Record;
+import org.xbill.DNS.Resolver;
+import org.xbill.DNS.Section;
+import org.xbill.DNS.SetResponse;
+import org.xbill.DNS.Type;
+
+/**
+ * LookupSession provides facilities to make DNS Queries. A LookupSession is intended to be long
+ * lived, and it's behaviour can be configured using the properties of the LookupSessionBuilder
+ * instance returned by the builder() method.
+ */
+@Builder
+public class LookupSession {
+  // remember to update javadoc if this is changed for the following two properties.
+  public static final int DEFAULT_MAX_ITERATIONS = 16;
+  public static final int DEFAULT_NDOTS = 1;
+  /** The {@link Resolver} to use to look up records. */
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}. Defaults to {@code 16}.
```

Comment:
I got the `{@value VALUE}` construct to work. (It turns out that with the latest released java 8 javadoc tool, one need to have the fully qualified field name of the constant field, presumably since the javadoc field is copied to an inner class by lombok).

I agree that it makes sense to add reading of the default to the code that constructs the sensible default instance, along with setting up caches and so on.